iceoryx2-bb/posix/tests/adaptive_wait_tests.rs

```diff
@@ -20,19 +20,48 @@ use std::time::Instant;
 const TIMEOUT: Duration = Duration::from_millis(50);
 
 #[test]
-fn adaptive_wait_wait_loop_initial_repetitions_plus_one_is_at_least_final_waiting_time() {
-    let start = Instant::now();
+fn adaptive_wait_wait_at_different_time_depends_on_repetition_times() {
     let mut counter: u64 = 0;
 
-    AdaptiveWaitBuilder::new()
-        .create()
-        .unwrap()
+    let mut waiter = AdaptiveWaitBuilder::new().create().unwrap();
+
+    waiter
+        .wait_while(move || -> bool {
+            counter += 1;
+            counter <= ADAPTIVE_WAIT_YIELD_REPETITIONS
+        })
+        .expect("failed to test wait_loop");
+    // the waiter starts to sleep ADAPTIVE_WAIT_INITIAL_WAITING_TIME instead of yield later.
+    assert_that!(waiter.yield_count(), eq ADAPTIVE_WAIT_YIELD_REPETITIONS);
+
+    // test sleep time ADAPTIVE_WAIT_INITIAL_WAITING_TIME
+    let start = Instant::now();
+    waiter
         .wait_while(move || -> bool {
             counter += 1;
-            counter < ADAPTIVE_WAIT_INITIAL_REPETITIONS
+            counter <= 1 // stop at the second turn
         })
         .expect("failed to test wait_loop");
+    assert_that!(start.elapsed(), ge ADAPTIVE_WAIT_INITIAL_WAITING_TIME);
 
+    waiter
+        .wait_while(move || -> bool {
+            counter += 1;
+            counter <= 99 // continue 99 turn to reach the edge of sleeping time.
+        })
+        .expect("failed to test wait_loop");
+
+    // verify the waiter will enter the next stage at the next repetition.
+    // the waiter starts to sleep longer as ADAPTIVE_WAIT_FINAL_WAITING_TIME
+    // instead of ADAPTIVE_WAIT_INITIAL_WAITING_TIME later.
+    assert_that!(waiter.yield_count(), eq ADAPTIVE_WAIT_INITIAL_REPETITIONS);
+    let start = Instant::now();
+    waiter
+        .wait_while(move || -> bool {
+            counter += 1;
+            counter <= 1 // stop at the second turn
+        })
+        .expect("failed to test wait_loop");
     assert_that!(start.elapsed(), ge ADAPTIVE_WAIT_FINAL_WAITING_TIME);
```

Comment:
Could you use here `assert_that!(start.elapsed(), time_at_least ADAPTIVE_WAIT_FINAL_WAITING_TIME);` The reason is that timing tests are notoriously unreliable.

Even when one just verifies at least timings, meaning that when one expects the process to sleep for time X it shall at least block for X. In Windows, for instance, this does not work - there are a lot of calls that have spurious wakeups, and then you have flaky tests all over the place. With `time_at_least` we cover exactly this and under the hood we add a variance. For Windows it is 0.0, meaning that we expect nothing from it and it is alright when it wakes up immediately (something to fix in the future inside the `iceoryx2-pal-posix` windows layer).

Linux and other RTOS have here a variance of 1.0, meaning it must satisfy the timing at least test.

Reply:
oops, sorry for missing this. Something new to me! I have changed all related usages in my PR. thanks a lot.

by the way, could I know how do you find this kind of nuance knowledge? It's quite system-dependent and looks like a experience-based knowledge.

Reply:
It is. We need to write a lot more documentation for iceoryx2 to capture all of this but we are lacking currently a bit behind. The highest priority has feature parity with good old iceoryx, then we have to provide a solid and detailed user documentation, capturing there all nuances and then we can go for in depth developer documentation.

Until then you have us and we try to pass as much experience onto you as possible :)

Reply:
sure, thanks for your kindly help:)